spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -0,0 +1,584 @@
+    // in the proposed ordering
+    // e.g. ORDERED -> UNORDERED, ORDERED -> DAG
+    abortTransactionUnless(
+        currentChannel.ordering.subsetOf(proposedUpgradeChannel.ordering)
+    )
+
+    // verify proofs of counterparty state
+    abortTransactionUnless(verifyChannelState(currentChannel, proofHeight, proofChannel, currentChannel.counterpartyChannelIdentifier, counterpartyChannel))
+    abortTransactionUnless(verifyUpgradeTimeout(currentChannel, proofHeight, proofUpgradeTimeout, currentChannel.counterpartyChannelIdentifier, upgradeTimeout))
+
+    if currentChannel.state == UPGRADE_INIT {
+        // if there is a crossing hello, ie an UpgradeInit has been called on both channelEnds,
+        // then we must ensure that the proposedUpgrade by the counterparty is the same as the currentChannel
+        // except for the channel state (upgrade channel will be in UPGRADE_TRY and current channel will be in UPGRADE_INIT)
+        // if the proposed upgrades on either side are incompatible, then we will restore the channel and cancel the upgrade.
+        currentChannel.state = UPGRADE_TRY
+        if !currentChannel.IsEqual(proposedUpgradeChannel) {
+            restoreChannel()
+            return
+        }
+    } else if currentChannel.state == OPEN {
+        // this is first message in upgrade handshake on this chain so we must store original channel in restore path
+        // in case we need to restore channel later.
+        privateStore.set(restorePath(portIdentifier, channelIdentifier), currentChannel)
+    } else {
+        // abort transaction if current channel is not in state: INIT or OPEN
+        abortTransactionUnless(false)
+    }
+
+    // either timeout height or timestamp must be non-zero
+    // if the upgrade feature is implemented on the TRY chain, then a relayer may submit a TRY transaction after the timeout.
+    // this will restore the channel on the executing chain and allow counterparty to use the CancelUpgradeMsg to restore their channel.
+    if timeoutHeight == 0 && timeoutTimestamp == 0 {
+        restoreChannel()
+        return
+    }
+    upgradeTimeout = UpgradeTimeout{
+        timeoutHeight: timeoutHeight,
+        timeoutTimestamp: timeoutTimestamp,
+    }
+
+    // counterparty-specified timeout must not have exceeded
+    if (currentHeight() > timeoutHeight && timeoutHeight != 0) ||
+        (currentTimestamp() > timeoutTimestamp && timeoutTimestamp != 0) {
+        restoreChannel()
+        return
+    }
+
+    // both channel ends must be mutually compatible.
+    // this function has been left unspecified since it will depend on the specific structure of the new channel.
+    // It is the responsibility of implementations to make sure that verification that the proposed new channels
+    // on either side are correctly constructed according to the new version selected.
+    if !IsCompatible(counterpartyChannel, proposedUpgradeChannel) {
+        restoreChannel()
+        return
+    }
+
```

Comment:
Sorry my mistake. Needed to define and prove counterpartyChannel earlier. counterpartyChannel is the channel that initiator upgraded to.

proposedUpgradeChannel is the channel that TRY is going to upgrade to.

Similar to the connection upgrade channel there needs to be some check on compatibility that will not be specified here since it is upgrade-specific.